### PR TITLE
Rectify: _normalize_subtype Encapsulation Leak

### LINKED ISSUE #1181: Fix: _normalize_subtype checks jsonl_context_exhausted instead of _is_context_exhausted()
## Problem

`_normalize_subtype()` in `session.py:925` checks the narrow `session.jsonl_context_exhausted` field (Path 1 only — JSONL flat-record scan) instead of calling `session._is_context_exhausted()` (the full 3-path method). This causes context exhaustion to be misclassified as `missing_completion_marker` when the exhaustion is detected via Path 3 (`is_error=True AND marker in result text`) rather than Path 1 (`jsonl_context_exhausted` flag from flat-record JSONL scan).

### Detection Architecture

`_is_context_exhausted()` has three detection paths (session.py:106-136):

1. **Path 1** — JSONL flat-record flag: `self.jsonl_context_exhausted` (set during `parse_session_result()` inline scan at line 401-412, fires only for records WITHOUT a `"message"` wrapper)
2. **Path 2** — Structured errors: `is_error=True AND marker in errors list`
3. **Path 3** — Result text: `is_error=True AND subtype in (SUCCESS, ERROR_MAX_TURNS) AND marker in result.lower()`

The `needs_retry` property (line 162-168) correctly delegates to `_is_context_exhausted()`, so **routing is always correct** (`retry_reason="resume"`). But `_normalize_subtype` line 925 only checks Path 1:

```python
# session.py:925 — only checks Path 1
if session.jsonl_context_exhausted:
    return "context_exhausted"
```

### Observed Failure

Session `a268d30c-3938-45d5-b145-07a583b357c9` (make-plan for issue #1164) hit the 200k context window limit. Claude Code emitted:

1. A **wrapped** synthetic assistant message: `{"type": "assistant", "message": {"model": "<synthetic>", "content": [{"text": "Prompt is too long"}], ...}, "error": "invalid_request"}`
2. A result record: `{"type": "result", "subtype": "success", "is_error": true, "result": "Prompt is too long", "terminal_reason": "blocking_limit"}`

Because the synthetic message has a `"message"` wrapper, the flat-record scan at line 401 (`elif "message" not in obj:`) skipped it. `jsonl_context_exhausted` stayed `False`.

However, `_is_context_exhausted()` Path 3 correctly fires: `is_error=True AND subtype=SUCCESS AND "prompt is too long" in result.lower()` → True.

Result: `needs_retry=True, retry_reason="resume"` (correct routing), but `_normalize_subtype` returns `"missing_completion_marker"` instead of `"context_exhausted"` (wrong diagnostic label).

### Impact

- **Routing**: Unaffected — orchestrator correctly interprets `retry_reason: "resume"` as context limit
- **Diagnostics**: `summary.json` records `subtype: "missing_completion_marker"` for what is actually a context exhaustion, making post-hoc analysis misleading
- **Audit queries**: Any query filtering on `subtype == "context_exhausted"` misses this session
- **Telemetry**: Context exhaustion frequency is undercounted

### Historical Context — Recurring Pattern

PR #420 (commit `0b89f9cb`) fixed the **same class of bug**: context exhaustion misclassified as `missing_completion_marker`. That fix added Path 1 (the `jsonl_context_exhausted` flag + flat-record JSONL scan) and updated `_normalize_subtype` line 925 to check `session.jsonl_context_exhausted`. But the fix was **incomplete**: it addressed the specific message format encountered (flat records without `"message"` wrapper) without ensuring `_normalize_subtype` respects ALL detection paths. The current session hit a different format (wrapped synthetic record with `is_error=True` in the result record), triggering Path 3 but not Path 1.

## Fix

One-line change in `session.py:925`:

```python
# Before:
if session.jsonl_context_exhausted:
    return "context_exhausted"

# After:
if session._is_context_exhausted():
    return "context_exhausted"
```

This makes `_normalize_subtype` delegate to the same authoritative method that `needs_retry` already uses, closing the abstraction leak.

## Test Gap

No test verifies that `_normalize_subtype` produces `"context_exhausted"` when context exhaustion is detected via Path 3 (is_error + result text) rather than Path 1 (jsonl flat-record flag). A test should construct a `ClaudeSessionResult` with `is_error=True`, `result="Prompt is too long"`, `jsonl_context_exhausted=False`, verify `_is_context_exhausted()` returns True, and assert `_normalize_subtype` returns `"context_exhausted"`.

## Acceptance Criteria

- [ ] `_normalize_subtype` line 925 calls `session._is_context_exhausted()` instead of checking `session.jsonl_context_exhausted`
- [ ] Test: Path 3 context exhaustion → `_normalize_subtype` returns `"context_exhausted"` (not `"missing_completion_marker"`)
- [ ] Test: Path 1 context exhaustion continues to work unchanged (regression guard)
- [ ] `task test-check` passes

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

`_normalize_subtype()` was a free function in `session.py` that read `jsonl_context_exhausted` directly instead of calling `session._is_context_exhausted()` — the canonical 3-path detection method. This caused context exhaustion detected via Path 2 (errors list) or Path 3 (result text) to be silently misclassified as `missing_completion_marker`.

The fix moves `_normalize_subtype` into a method on `ClaudeSessionResult` (`normalize_subtype(self, outcome, completion_marker) -> str`), which delegates to `self._is_context_exhausted()` — making all three detection paths visible and the abstraction leak structurally impossible. Two secondary issues are also addressed: removal of the unreachable dead function `_detect_flat_context_exhaustion()`, and a detection gap in `_llm_triage.py` that missed the `jsonl_context_exhausted=True, is_error=False` case from Path 1.

### Module Dependency Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 50, 'rankSpacing': 70, 'curve': 'basis'}}}%%
graph TB
    %% CLASS DEFINITIONS %%
    classDef cli fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef integration fill:#c62828,stroke:#ef9a9a,stroke-width:2px,color:#fff;
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;

    subgraph AppLayer ["APP LAYER — Top-level modules"]
        direction LR
        LLM_TRIAGE["● _llm_triage.py<br/>━━━━━━━━━━<br/>Contract staleness triage<br/>Spawns Haiku subprocess<br/>uses parse_session_result"]
    end

    subgraph L2Layer ["L2 — Recipe"]
        direction LR
        RECIPE["autoskillit.recipe<br/>━━━━━━━━━━<br/>StaleItem, load_bundled_manifest<br/>(unchanged)"]
    end

    subgraph L1Exec ["L1 — execution/ package"]
        direction TB
        HEADLESS["● execution/headless.py<br/>━━━━━━━━━━<br/>_build_skill_result()<br/>_execute_claude_headless()<br/>run_headless_core()<br/>DefaultHeadlessExecutor<br/>calls session.normalize_subtype()"]

        SESSION["● execution/session.py<br/>━━━━━━━━━━<br/>ClaudeSessionResult (dataclass)<br/>+ normalize_subtype() ← moved here<br/>parse_session_result()<br/>_compute_outcome()<br/>_compute_success()<br/>ContentState, FAILURE_SUBTYPES"]

        EXEC_PEERS["execution peers<br/>━━━━━━━━━━<br/>_headless_scan.py<br/>clone_guard.py<br/>commands.py<br/>process.py<br/>recording.py<br/>session_log.py"]
    end

    subgraph L1Workspace ["L1 — workspace/"]
        direction LR
        WORKSPACE["autoskillit.workspace<br/>━━━━━━━━━━<br/>bundled_skills_dir()<br/>(unchanged)"]
    end

    subgraph L0Layer ["L0 — autoskillit.core (FOUNDATION — zero autoskillit imports)"]
        direction LR
        CORE["autoskillit.core<br/>━━━━━━━━━━<br/>CliSubtype, SessionOutcome<br/>RetryReason, TerminationReason<br/>ChannelConfirmation, SkillResult<br/>CONTEXT_EXHAUSTION_MARKER<br/>get_logger, truncate_text"]
    end

    subgraph Tests ["TESTS (modified)"]
        direction LR
        TEST_NORM["● test_normalize_subtype.py<br/>━━━━━━━━━━<br/>tests ClaudeSessionResult<br/>.normalize_subtype() directly"]
        TEST_TOOLS["● test_tools_execution.py<br/>━━━━━━━━━━<br/>integration: expects<br/>context_exhausted subtype"]
        TEST_ARCH["● test_subpackage_isolation.py<br/>━━━━━━━━━━<br/>verifies session.py imports<br/>only from L0 (core)"]
    end

    %% VALID DEPENDENCIES (downward) %%
    LLM_TRIAGE -->|"parse_session_result<br/>run_managed_async"| SESSION
    LLM_TRIAGE -->|"StaleItem<br/>load_bundled_manifest"| RECIPE
    LLM_TRIAGE -->|"ClaudeFlags, OutputFormat<br/>SubprocessResult, build_claude_env"| CORE
    LLM_TRIAGE -->|"bundled_skills_dir()"| WORKSPACE

    HEADLESS -->|"ClaudeSessionResult<br/>_compute_outcome<br/>_compute_success<br/>_check_expected_patterns<br/>_truncate, parse_session_result"| SESSION
    HEADLESS -->|"CAMPAIGN_ID_ENV_VAR<br/>ChannelConfirmation, CliSubtype<br/>SkillResult, RetryReason, etc."| CORE
    HEADLESS -->|"_scan_jsonl_write_paths"| EXEC_PEERS
    HEADLESS -->|"build_food_truck_cmd<br/>build_headless_resume_cmd<br/>build_leaf_headless_cmd"| EXEC_PEERS
    HEADLESS -->|"_marker_is_standalone"| EXEC_PEERS
    HEADLESS -->|"RecordingSubprocessRunner"| EXEC_PEERS
    HEADLESS -->|"flush_session_log (deferred)"| EXEC_PEERS

    SESSION -->|"CliSubtype, SessionOutcome<br/>RetryReason, TerminationReason<br/>ChannelConfirmation, SkillResult<br/>CONTEXT_EXHAUSTION_MARKER<br/>get_logger, truncate_text"| CORE

    RECIPE --> CORE
    WORKSPACE --> CORE
    EXEC_PEERS --> CORE

    %% Test imports %%
    TEST_NORM -.->|"tests"| SESSION
    TEST_TOOLS -.->|"integration tests"| HEADLESS
    TEST_ARCH -.->|"import isolation check"| SESSION

    %% CLASS ASSIGNMENTS %%
    class LLM_TRIAGE cli;
    class SESSION stateNode;
    class HEADLESS handler;
    class EXEC_PEERS phase;
    class RECIPE output;
    class WORKSPACE output;
    class CORE integration;
    class TEST_NORM,TEST_TOOLS,TEST_ARCH detector;
```

### Process Flow Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 40, 'rankSpacing': 50, 'curve': 'basis'}}}%%
flowchart TB
    %% CLASS DEFINITIONS %%
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;

    %% TERMINALS %%
    START([START: Claude CLI stdout])
    COMPLETE([COMPLETE: SkillResult returned])
    ERROR([ERROR: crashed / cancelled])

    subgraph Scan ["NDJSON Scan — parse_session_result()"]
        direction TB
        LineLoop["Iterate stdout lines<br/>━━━━━━━━━━<br/>json.loads each line"]
        RecordType{"record type?"}
        ResultRecord["acc.result_obj = obj<br/>━━━━━━━━━━<br/>type=result → authoritative"]
        AssistantWrapped["Wrapped assistant record<br/>━━━━━━━━━━<br/>collect tool_uses,<br/>assistant_messages,<br/>stop_reasons"]
        AssistantFlat["Flat assistant record<br/>━━━━━━━━━━<br/>no 'message' wrapper"]
        FlatCheck{"output_tokens==0<br/>AND marker in<br/>content[].text?"}
        SetJSONL["acc.jsonl_context_exhausted = True<br/>━━━━━━━━━━<br/>Path 1 signal captured"]
    end

    subgraph Construct ["Session Construction"]
        direction TB
        HasResult{"acc.result_obj<br/>present?"}
        BuildFromResult["Extract subtype, is_error,<br/>result_text, errors<br/>━━━━━━━━━━<br/>reads from acc.result_obj"]
        BuildFallback{"acc.jsonl_context_exhausted?"}
        FallbackCE["subtype = CONTEXT_EXHAUSTION<br/>is_error = True"]
        FallbackUP["subtype = UNPARSEABLE<br/>is_error = True"]
        BuildCSR["● ClaudeSessionResult<br/>━━━━━━━━━━<br/>subtype, is_error, result,<br/>errors, jsonl_context_exhausted,<br/>assistant_messages, tool_uses"]
    end

    subgraph Recovery ["Recovery Gates — _build_skill_result()"]
        direction TB
        ChannelBGate{"Channel B +<br/>recoverable subtype?"}
        SeparateMarker["_recover_from_separate_marker()<br/>━━━━━━━━━━<br/>rebuild result from<br/>assistant_messages"]
        PatternRecovery["_recover_block_from_assistant_messages()<br/>━━━━━━━━━━<br/>find patterns in<br/>assistant_messages"]
        Synthesis["_synthesize_from_write_artifacts()<br/>━━━━━━━━━━<br/>inject path token<br/>from Write/Edit calls"]
    end

    subgraph Outcome ["Outcome Computation"]
        direction TB
        ComputeOutcome["● _compute_outcome()<br/>━━━━━━━━━━<br/>_compute_success() +<br/>_compute_retry() +<br/>contradiction guard +<br/>dead-end guard"]
        OutcomeResult{"SessionOutcome"}
        OS_SUCCEEDED["SUCCEEDED"]
        OS_RETRIABLE["RETRIABLE"]
        OS_FAILED["FAILED"]
    end

    subgraph Normalize ["● Subtype Normalization — normalize_subtype()"]
        direction TB
        SubtypeSwitch{"session.subtype?"}
        PassthroughSuccess["return 'success'<br/>━━━━━━━━━━<br/>SUCCEEDED + SUCCESS<br/>→ passthrough"]
        DownwardNorm["Downward normalization<br/>━━━━━━━━━━<br/>non-SUCCEEDED + SUCCESS<br/>→ synthesize label"]
        EmptyCheck{"result.strip()<br/>empty?"}
        EmptyResult["return 'empty_result'"]
        CtxExhausted{"● _is_context_exhausted()?<br/>━━━━━━━━━━<br/>delegates to method<br/>(was: direct flag check)"}
        CtxLabel["return 'context_exhausted'"]
        MarkerCheck{"completion_marker<br/>not in result?"}
        MissingMarker["return 'missing_completion_marker'"]
        AdjFail["return 'adjudicated_failure'"]
        FailureSubtypes["Failure subtypes<br/>(UNKNOWN, EMPTY_OUTPUT,<br/>UNPARSEABLE, TIMEOUT,<br/>IDLE_STALL)<br/>━━━━━━━━━━<br/>upward normalize if SUCCEEDED"]
        PassthroughOther["Passthrough subtypes<br/>(ERROR_MAX_TURNS,<br/>ERROR_DURING_EXECUTION,<br/>CONTEXT_EXHAUSTION,<br/>INTERRUPTED)<br/>━━━━━━━━━━<br/>return self.subtype"]
    end

    subgraph CtxDetect ["● _is_context_exhausted() — 3-Path Detection"]
        direction TB
        Path1{"Path 1:<br/>jsonl_context_exhausted?"}
        Path1True["return True<br/>━━━━━━━━━━<br/>JSONL flat-record<br/>race-resilient bypass"]
        IsErrorGate{"is_error?"}
        Path2{"Path 2:<br/>marker in errors[]?"}
        Path2True["return True<br/>━━━━━━━━━━<br/>structured errors list"]
        Path3{"Path 3:<br/>subtype in (SUCCESS,<br/>ERROR_MAX_TURNS)<br/>AND marker in result?"}
        Path3True["return True<br/>━━━━━━━━━━<br/>result text fallback"]
        ReturnFalse["return False"]
    end

    %% FLOW %%
    START --> LineLoop
    LineLoop --> RecordType
    RecordType -->|"type=result"| ResultRecord
    RecordType -->|"type=assistant<br/>+ message wrapper"| AssistantWrapped
    RecordType -->|"type=assistant<br/>no message key"| AssistantFlat
    AssistantFlat --> FlatCheck
    FlatCheck -->|"yes"| SetJSONL
    FlatCheck -->|"no"| LineLoop
    SetJSONL --> LineLoop
    ResultRecord --> LineLoop
    AssistantWrapped --> LineLoop
    LineLoop -->|"all lines done"| HasResult

    HasResult -->|"yes"| BuildFromResult
    HasResult -->|"no"| BuildFallback
    BuildFallback -->|"yes"| FallbackCE
    BuildFallback -->|"no"| FallbackUP
    BuildFromResult --> BuildCSR
    FallbackCE --> BuildCSR
    FallbackUP --> BuildCSR

    BuildCSR --> ChannelBGate
    ChannelBGate -->|"yes"| SeparateMarker
    ChannelBGate -->|"no / not applicable"| PatternRecovery
    SeparateMarker --> PatternRecovery
    PatternRecovery --> Synthesis
    Synthesis --> ComputeOutcome

    ComputeOutcome --> OutcomeResult
    OutcomeResult --> OS_SUCCEEDED
    OutcomeResult --> OS_RETRIABLE
    OutcomeResult --> OS_FAILED

    OS_SUCCEEDED --> SubtypeSwitch
    OS_RETRIABLE --> SubtypeSwitch
    OS_FAILED --> SubtypeSwitch

    SubtypeSwitch -->|"SUCCESS"| PassthroughSuccess
    PassthroughSuccess -->|"SUCCEEDED"| COMPLETE
    SubtypeSwitch -->|"SUCCESS +<br/>non-SUCCEEDED"| DownwardNorm
    DownwardNorm --> EmptyCheck
    EmptyCheck -->|"yes"| EmptyResult
    EmptyCheck -->|"no"| CtxExhausted
    CtxExhausted -->|"yes"| CtxLabel
    CtxExhausted -->|"no"| MarkerCheck
    MarkerCheck -->|"yes"| MissingMarker
    MarkerCheck -->|"no"| AdjFail

    SubtypeSwitch -->|"failure subtypes"| FailureSubtypes
    FailureSubtypes -->|"SUCCEEDED → 'success'"| COMPLETE
    FailureSubtypes -->|"non-SUCCEEDED → passthrough"| COMPLETE
    SubtypeSwitch -->|"passthrough subtypes"| PassthroughOther
    PassthroughOther --> COMPLETE
    EmptyResult --> COMPLETE
    CtxLabel --> COMPLETE
    MissingMarker --> COMPLETE
    AdjFail --> COMPLETE

    CtxExhausted --> Path1
    Path1 -->|"yes"| Path1True
    Path1 -->|"no"| IsErrorGate
    Path1True -->|"return True"| CtxLabel
    IsErrorGate -->|"no"| ReturnFalse
    IsErrorGate -->|"yes"| Path2
    Path2 -->|"yes"| Path2True
    Path2 -->|"no"| Path3
    Path2True -->|"return True"| CtxLabel
    Path3 -->|"yes"| Path3True
    Path3 -->|"no"| ReturnFalse
    Path3True -->|"return True"| CtxLabel
    ReturnFalse -->|"return False"| MarkerCheck

    START -->|"runner crashes"| ERROR

    %% CLASS ASSIGNMENTS %%
    class START,COMPLETE,ERROR terminal;
    class LineLoop,AssistantWrapped,AssistantFlat,BuildFromResult,FallbackCE,FallbackUP handler;
    class RecordType,HasResult,BuildFallback,FlatCheck,ChannelBGate,OutcomeResult phase;
    class BuildCSR,SetJSONL stateNode;
    class SeparateMarker,PatternRecovery,Synthesis,ComputeOutcome handler;
    class OS_SUCCEEDED,OS_RETRIABLE,OS_FAILED stateNode;
    class SubtypeSwitch,EmptyCheck,MarkerCheck phase;
    class DownwardNorm,PassthroughSuccess,FailureSubtypes,PassthroughOther handler;
    class CtxExhausted,Path1,IsErrorGate,Path2,Path3 detector;
    class EmptyResult,CtxLabel,MissingMarker,AdjFail output;
    class Path1True,Path2True,Path3True,ReturnFalse output;
    class ResultRecord stateNode;
```

### Error/Resilience Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 40, 'rankSpacing': 50, 'curve': 'basis'}}}%%
flowchart TB
    %% CLASS DEFINITIONS %%
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef gap fill:#ff6f00,stroke:#ffa726,stroke-width:2px,color:#000;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;

    %% ── ENTRY POINTS ────────────────────────────────────────────────────── %%
    NDJSON["Raw NDJSON stdout<br/>━━━━━━━━━━<br/>Claude CLI output stream"]
    TRIAGE_IN["LLM Triage Request<br/>━━━━━━━━━━<br/>hash_mismatch batch"]

    %% ── PARSING LAYER ───────────────────────────────────────────────────── %%
    subgraph Parsing ["PARSING (● execution/session.py)"]
        direction TB
        SCAN["_ParseAccumulator scan<br/>━━━━━━━━━━<br/>Accumulates: tool_uses,<br/>assistant_messages,<br/>stop_reasons"]
        FLAT["Flat-record detection<br/>━━━━━━━━━━<br/>output_tokens==0 + marker<br/>→ jsonl_context_exhausted=True"]
        CSR["● ClaudeSessionResult<br/>━━━━━━━━━━<br/>subtype, is_error, result,<br/>errors, jsonl_context_exhausted"]
    end

    %% ── CONTEXT EXHAUSTION DETECTION ────────────────────────────────────── %%
    subgraph CtxDetection ["CONTEXT EXHAUSTION DETECTION (● session.py)"]
        direction TB
        P1["Path 1: JSONL flat-record<br/>━━━━━━━━━━<br/>jsonl_context_exhausted=True<br/>bypasses is_error guard"]
        P2["Path 2: errors list<br/>━━━━━━━━━━<br/>is_error=True AND<br/>marker in errors[]"]
        P3["Path 3: result text<br/>━━━━━━━━━━<br/>is_error=True AND marker in<br/>result (subtype=success only)"]
        ICE{"_is_context_exhausted()<br/>returns True?"}
    end

    %% ── SUBTYPE NORMALIZATION GATE (PR FOCUS) ───────────────────────────── %%
    subgraph NormGate ["● normalize_subtype() — MOVED TO METHOD (PR FIX)"]
        direction TB
        CLS1{"Class 1:<br/>success subtype +<br/>non-SUCCEEDED?"}
        CLS2{"Class 2:<br/>failure subtype +<br/>SUCCEEDED?"}
        EMPTY_R["→ 'empty_result'<br/>━━━━━━━━━━<br/>result is blank"]
        CTX_EX["→ 'context_exhausted'<br/>━━━━━━━━━━<br/>delegates _is_context_exhausted()<br/>(bug fix: was jsonl_context_exhausted)"]
        MISS_M["→ 'missing_completion_marker'<br/>━━━━━━━━━━<br/>non-empty result, marker absent"]
        ADJ_F["→ 'adjudicated_failure'<br/>━━━━━━━━━━<br/>catch-all downward label"]
        UP_OK["→ 'success'<br/>━━━━━━━━━━<br/>upward normalization"]
        PASS["→ passthrough<br/>━━━━━━━━━━<br/>error_max_turns,<br/>context_exhaustion,<br/>interrupted, etc."]
    end

    %% ── OUTCOME COMPUTATION ─────────────────────────────────────────────── %%
    subgraph Outcome ["OUTCOME COMPUTATION (session.py)"]
        direction TB
        CS["_compute_success()<br/>━━━━━━━━━━<br/>Channel B bypass gate<br/>+ termination dispatch"]
        CR["_compute_retry()<br/>━━━━━━━━━━<br/>API-level (needs_retry)<br/>+ termination dispatch"]
        CONTR{"Contradiction guard:<br/>success AND needs_retry?"}
        DEAD{"Dead-end guard:<br/>!success AND !retry<br/>+ channel confirmed?"}
        ABSENT["ContentState.ABSENT<br/>━━━━━━━━━━<br/>drain-race → RETRIABLE"]
        TERM_FAIL["Terminal failure<br/>━━━━━━━━━━<br/>CONTRACT_VIOLATION /<br/>SESSION_ERROR → FAILED"]
    end

    %% ── RECOVERY MECHANISMS (headless.py) ──────────────────────────────── %%
    subgraph Recovery ["RECOVERY MECHANISMS (● execution/headless.py)"]
        direction TB
        CB_REC["Channel B drain-race<br/>recovery<br/>━━━━━━━━━━<br/>UNPARSEABLE/EMPTY_OUTPUT<br/>+ marker in assistant_msgs"]
        PAT_REC["Pattern recovery<br/>━━━━━━━━━━<br/>expected_output_patterns<br/>missing → scan assistant_msgs"]
        SYNTH["Artifact synthesis<br/>━━━━━━━━━━<br/>UNMONITORED only:<br/>Write/Edit tool_use paths"]
        NUDGE["Contract nudge<br/>━━━━━━━━━━<br/>CONTRACT_RECOVERY:<br/>60s resume session"]
        BG{"Budget guard<br/>━━━━━━━━━━<br/>consecutive > max?"}
        ZWG{"Zero-write gate<br/>━━━━━━━━━━<br/>write_expected<br/>AND writes==0?"}
        CRG{"CONTRACT_RECOVERY<br/>gate<br/>━━━━━━━━━━<br/>adjudicated_failure<br/>+ write evidence?"}
    end

    %% ── LLM TRIAGE ERROR HANDLING (● _llm_triage.py) ───────────────────── %%
    subgraph Triage ["● LLM TRIAGE (_llm_triage.py)"]
        direction TB
        HAIKU["Haiku subprocess<br/>━━━━━━━━━━<br/>claude -p batch prompt<br/>30s timeout"]
        TM_ERR["TimeoutError / OSError<br/>━━━━━━━━━━<br/>→ all meaningful=True<br/>(fail-safe)"]
        PARSE_ERR["session.is_error<br/>OR needs_retry<br/>━━━━━━━━━━<br/>→ batch parse failed<br/>all meaningful=True"]
        LEN_ERR["JSON length mismatch<br/>━━━━━━━━━━<br/>→ fallback: all meaningful=True"]
        ITEM_ERR["Per-item key error<br/>━━━━━━━━━━<br/>→ that item meaningful=True<br/>siblings preserved"]
    end

    %% ── TERMINAL STATES ─────────────────────────────────────────────────── %%
    T_SUCCESS([SUCCEEDED<br/>SkillResult success=True])
    T_RETRY([RETRIABLE<br/>needs_retry=True])
    T_FAILED([FAILED<br/>success=False needs_retry=False])
    T_BUDGET([BUDGET_EXHAUSTED<br/>retry capped])
    T_TRIAGE([TRIAGE RESULT<br/>meaningful + summary list])

    %% ── CONNECTIONS ─────────────────────────────────────────────────────── %%

    %% Parsing flow
    NDJSON --> SCAN
    SCAN -->|"flat assistant record"| FLAT
    FLAT -->|"jsonl_context_exhausted=True"| CSR
    SCAN --> CSR

    %% Context exhaustion detection
    CSR --> P1
    CSR --> P2
    CSR --> P3
    P1 -->|"True (priority 1)"| ICE
    P2 -->|"True (priority 2)"| ICE
    P3 -->|"True (priority 3)"| ICE

    %% Normalize subtype
    ICE -->|"feeds into"| CLS1
    CSR --> CLS1
    CSR --> CLS2
    CLS1 -->|"empty result"| EMPTY_R
    CLS1 -->|"_is_context_exhausted()"| CTX_EX
    CLS1 -->|"marker absent"| MISS_M
    CLS1 -->|"catch-all"| ADJ_F
    CLS2 -->|"outcome=SUCCEEDED"| UP_OK
    CLS2 -->|"not success subtype"| PASS

    %% Outcome computation
    CSR --> CS
    CSR --> CR
    CS --> CONTR
    CR --> CONTR
    CONTR -->|"yes: demote success"| DEAD
    CONTR -->|"no contradiction"| DEAD
    DEAD -->|"ABSENT state"| ABSENT
    DEAD -->|"terminal"| TERM_FAIL
    ABSENT --> T_RETRY
    TERM_FAIL --> T_FAILED

    %% Recovery chain in headless.py
    CSR --> CB_REC
    CB_REC -->|"recovered"| PAT_REC
    CB_REC -->|"not recovered"| PAT_REC
    PAT_REC -->|"recovered"| SYNTH
    PAT_REC -->|"not recovered"| SYNTH
    SYNTH --> CRG
    CRG -->|"yes: promote RETRIABLE"| NUDGE
    CRG -->|"no"| ZWG
    NUDGE -->|"nudge success"| T_SUCCESS
    NUDGE -->|"nudge fail"| T_RETRY
    ZWG -->|"zero_writes triggered"| T_RETRY
    ZWG -->|"ok"| BG
    BG -->|"budget exceeded"| T_BUDGET
    BG -->|"ok"| T_SUCCESS
    T_BUDGET --> T_FAILED

    %% Triage flow
    TRIAGE_IN --> HAIKU
    HAIKU -->|"TimeoutError/OSError"| TM_ERR
    HAIKU -->|"session parsed"| PARSE_ERR
    PARSE_ERR -->|"is_error or needs_retry"| LEN_ERR
    PARSE_ERR -->|"valid"| LEN_ERR
    LEN_ERR -->|"mismatch"| TM_ERR
    LEN_ERR -->|"ok"| ITEM_ERR
    TM_ERR --> T_TRIAGE
    ITEM_ERR --> T_TRIAGE

    %% CLASS ASSIGNMENTS %%
    class NDJSON,TRIAGE_IN stateNode;
    class SCAN,FLAT handler;
    class CSR stateNode;
    class P1,P2,P3 detector;
    class ICE detector;
    class CLS1,CLS2 phase;
    class EMPTY_R,MISS_M,ADJ_F gap;
    class CTX_EX output;
    class UP_OK,PASS output;
    class CS,CR handler;
    class CONTR,DEAD phase;
    class ABSENT gap;
    class TERM_FAIL gap;
    class CB_REC,PAT_REC,SYNTH,NUDGE output;
    class BG,ZWG,CRG phase;
    class HAIKU handler;
    class TM_ERR,PARSE_ERR,LEN_ERR,ITEM_ERR gap;
    class T_SUCCESS,T_RETRY,T_FAILED,T_BUDGET,T_TRIAGE terminal;
```

Closes #1181

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/remediation-20260424-134737-904362/.autoskillit/temp/rectify/rectify_normalize_subtype_encapsulation_2026-04-24_000000.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | uncached | output | cache_read | cache_write | count | time |
|------|----------|--------|------------|-------------|-------|------|
| investigate | 84 | 6.6k | 247.5k | 31.4k | 1 | 4m 3s |
| rectify | 105 | 13.9k | 491.6k | 53.7k | 1 | 7m 12s |
| dry_walkthrough | 150 | 12.5k | 968.2k | 60.4k | 1 | 3m 39s |
| implement | 278 | 16.9k | 2.3M | 78.2k | 1 | 4m 40s |
| assess | 184 | 6.7k | 938.4k | 45.9k | 1 | 8m 33s |
| prepare_pr | 60 | 5.0k | 197.5k | 25.9k | 1 | 1m 19s |
| run_arch_lenses | 192 | 19.4k | 946.0k | 195.1k | 3 | 5m 50s |
| compose_pr | 67 | 10.1k | 263.1k | 37.2k | 1 | 2m 26s |
| **Total** | 1.1k | 91.2k | 6.3M | 527.8k | | 37m 45s |

### CODE REVIEW

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

File: src/autoskillit/execution/session.py [hunk 1]
```diff
@@ -174,6 +174,60 @@ def retry_reason(self) -> RetryReason:
             return RetryReason.RESUME
         return RetryReason.NONE
 
+    def normalize_subtype(self, outcome: SessionOutcome, completion_marker: str) -> str:
+        """Normalize subtype against the adjudicated outcome to eliminate contradictions.
```
Comment (Trecek) [inline src/autoskillit/execution/session.py:178]: [warning] slop: Docstring for normalize_subtype repeats information already expressed by the match branches and type annotation. The Class 1/Class 2 labels and the Return-type-is-str paragraph are redundant — the code is self-explanatory. Trim to a single-line summary.

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: warning-only findings detected. See inline comments — no blocking changes required.

### ⚠️ Outside Diff Range

These findings target lines not in the diff and could not be posted as inline comments:

**tests/execution/test_normalize_subtype.py**
- **L200** [warning/tests]: No test covers the edge case where `jsonl_context_exhausted=True` AND `result` is empty. In `normalize_subtype`, the empty-result check fires before `_is_context_exhausted()`, so such a session returns `empty_result` instead of `context_exhausted`. Consider confirming whether this priority ordering is intended.

File: src/autoskillit/execution/session.py (hunk 1 above)
Comment (Trecek) [reply src/autoskillit/execution/session.py:178]: Investigated — this is intentional. The docstring contains content not expressible from the code alone: (1) the 'Exhaustive match over CliSubtype ensures mypy flags any unhandled member when the enum is extended' sentence explains WHY the exhaustive match is used — an architectural contract for future-proofing (see assert_never at line 229); (2) 'Return type is str (not CliSubtype) because downward normalization synthesizes labels that are SkillResult-level subtypes' explains a deliberate type-system deviation that would otherwise appear accidental. The Class 1/Class 2 label paragraphs are admittedly redundant, but trimming to a single-line summary would delete both valuable explanations. The reviewer's blanket slop classification over-generalizes.
